README
SignaturePad: A jQuery plugin for assisting in the creation of an HTML5 canvas
based signature pad. Records the drawn signature in JSON for later regeneration.

Dependencies: excanvas, json2, jquery-1.3.2+

@project    ca.thomasjbradley.applications.signaturepad
@author     Thomas J Bradley <hey@thomasjbradley.ca>
@link       http://thomasjbradley.ca/lab/signature-pad
@link       http://github.com/thomasjbradley/signature-pad
@copyright  Copyright MMXI, Thomas J Bradley
@license    New BSD License
@version    1.5.1

Versioned using Semantic Versioning <http://semver.org/>


COMPLETE DOCUMENTATION:
http://thomasjbradley.ca/lab/signature-pad

SOURCE CODE:
http://github.com/thomasjbradley/signature-pad


VERSION HISTORY:

1.5.1 (Feb. 23, 2011)
- Added an option/fixed a bug: the pen would be jagged if thicker than 4 pixels
- Fixed a bug: if the pen left the canvas it would continue to draw
- Switch to proper JSDoc
- Cleanup and removal of extraneous semi-colons

1.5.0 (Dec. 3, 2010)
- Added an option to disable field validation
- Cleaned up some code

1.4.0 (Nov. 30, 2010)
- Regenerate method now repopulates the output so signature can be appended
- Removed some extra debug code that got missed

1.3.2 (Oct. 31, 2010)
- Fixed a bug on iPhone created by the fix in v1.3.1:
  if signature pad was at the top iPhone may scroll upwards

1.3.1 (Oct. 30, 2010)
- Fixed a bug on iPad when zoomed in:
  the signature would draw in the wrong location

1.3.0 (Aug. 29, 2010)
- Added getSignatureImage() method that calls toDataURL() on the canvas

1.2.2 (Aug. 29, 2010)
- Fixed an offset bug

1.2.1 (Jun. 7, 2010)
- Added namespaces to all event bindings

1.2.0 (May 28, 2010)
- Added touch device support

1.1.3 (May 28, 2010)
- Fixed bug in validation if multiple signature pads were used in jQuery 1.4.2

1.1.2 (May 26, 2010)
- Fixed an XSS-like bug, where entering a <script> or an <iframe> element into
  the name field would cause the code to be executed
  < and > are now converted to entities

1.1.1 (Jan. 26, 2010)
- Fixed a bug in IE 8, where IE 8 would not regenerate signatures
  Required a new version of excanvas (r71 from svn trunk)
  IE6 still requires older release r3 to work

1.1.0 (Jan. 4, 2010)
- Added option to require and validate a drawn signature
- Added option to default to drawIt mode
- Fixed a bug where pressing validate multiple times would remove name input

1.0.1 (Dec. 3, 2009)
- Fixed @font-face support for Google Chrome by adding SVG fonts
- Fixed future @font-face support by adding WOFF font

1.0.0 (Nov. 30, 2009)
- Initial Release


CHANGELOG.md
# Changelog

**2.5.2 (Dec. 22, 2014)**

- Merged [Pull Request #124](https://github.com/thread-pond/signature-pad/issues/124) to fix a missing semi-colon at the end.

**2.5.1 (Apr. 15, 2014)**

- Fixed a bug in IE 8 related to `trim()`, [Issue #113](https://github.com/thread-pond/signature-pad/issues/113)

**2.5.0 (Oct. 16, 2013)**

- Added an API to retrieve the version programmatically, [Issue #96](https://github.com/thread-pond/signature-pad/issues/96)

**2.4.2 (Sep. 25, 2013)**

- Fixed a bug where onDrawEnd wasn’t firing properly on touch devices, [Issue #92](https://github.com/thread-pond/signature-pad/issues/92)
- Changed a CSS selector for better Windows 8 support, [Issue #91](https://github.com/thread-pond/signature-pad/issues/91)
- Fixed a bug where right clicking would cause a dot to be drawn, [Issue #80](https://github.com/thread-pond/signature-pad/issues/80)

**2.4.1 (Jul. 26, 2013)**

- Fixed a bug where onDrawEnd was firing too often and at incorrect times, [Issue #84](https://github.com/thread-pond/signature-pad/issues/84)

**2.4.0 (Jul. 16, 2013)**

- Added Grunt.js as the build tool
- Added an `onDraw` callback; [Pull Request #60](https://github.com/thread-pond/signature-pad/pull/60)
- Added an `onDrawEnd` callback; [Pull Request #70](https://github.com/thread-pond/signature-pad/pull/70)
- Added a new full window example; [Pull Request #49](https://github.com/thread-pond/signature-pad/pull/49)
- Updated the examples to be more modern; [Pull Request #49](https://github.com/thread-pond/signature-pad/pull/49)
- Improved the signature handling when leaving the `<canvas>`; [Issue #38](https://github.com/thread-pond/signature-pad/issues/38)
- Fixed some event and touch related bugs for: IE/10 with [Pull Request #49](https://github.com/thread-pond/signature-pad/pull/49); Android/4 with [Issue #78](https://github.com/thread-pond/signature-pad/issues/78); more browsers with [Issue #38](https://github.com/thread-pond/signature-pad/issues/38)
- Fixed the typed signature scaling bug [Issue #54](https://github.com/thread-pond/signature-pad/issues/54)
- Fixed up the .jshintrc file

**2.3.0 (Jun. 12, 2012)**

- Added the ability to get the API by re-executing signature pad on the element; [Issue #23](https://github.com/thread-pond/signature-pad/issues/23) and many more
- Added a function to allow updating of options later on
- Performance fix: if options.output is falsey the JSON will not be calculated; [Issue #15](https://github.com/thread-pond/signature-pad/issues/15)
- Removes focus from input on initialization to close any open keyboards
- Fixed [Issue #25](https://github.com/thread-pond/signature-pad/issues/25)
- Opened up the `validateForm()` function to allow for inclusion in other form validation code

**2.2.0 (Feb. 26, 2012)**

- Updated the custom cursor to be more browser compatible. Doesn’t work in IE < 8 because FlashCanvas overwrites the cursor. Thanks to: <http://www.useragentman.com/blog/2011/12/21/cross-browser-css-cursor-images-in-depth/>
- Updated CSS: fixed some glaring vendor prefix bugs, neutralized the colours, and updated the fonts

**2.1.2 (Feb. 21, 2012)**

- Bug fix: added `clearRect()` code to allow transparent backgrounds on the canvas—[Issue #18](https://github.com/thread-pond/signature-pad/issues/18)

**2.1.1 (Sep. 30, 2011)**

- Removed some extra, un-needed arguments from the `onBeforeValidate` and `onFormError` callbacks

**2.1.0 (Sep. 29, 2011)**

- Added Firefox Mobile support
- Added callbacks for validation: `onBeforeValidate`, `onFormError`—[Issue #7](https://github.com/thread-pond/signature-pad/issues/7)
- Changed `getSignatureImage()` to use a shadow canvas and render the signature without the line
- Cleaned up event creation and removal
- Fixed a bug with Playbook and other non-iOS Webkit browsers; caused by a bug fix for iOS browsers—[Issue #8](https://github.com/thread-pond/signature-pad/issues/8)
- Can now pass arguments to `getSignatureImage()`, which will be forwarded to `toDataURL()`—[Issue #9](https://github.com/thread-pond/signature-pad/issues/9)
- Fixed [Issue #10](https://github.com/thread-pond/signature-pad/issues/10)

**2.0.4 (May 12, 2011)**

- Fixed a bug with `defaultAction: drawIt`: the `.typeItDesc` was initially being displayed when it shouldn’t

**2.0.3 (May 10, 2011)**

- Fixed a bug with `bgColour`: the colour was not being honoured when clearing the canvas

**2.0.2 (Apr. 7, 2011)**

- Added an unintentional missing feature: when clicking on the pad, a dot is now drawn

**2.0.1 (Apr. 7, 2011)**

- Fixed a bug in `regenerate()` related to the `penCap`

**2.0.0 (Feb. 25, 2011)**

- Switched to FlashCanvas instead of ExplorerCanvas (more reliable, less hacks needed, actively being developed)
- `.getSignatureImage()` now works cross-browser thanks to FlashCanvas
- Fixed [Issue #2](https://github.com/thread-pond/signature-pad/issues/2)

**1.5.1 (Feb. 23, 2011)**

- Added an option/fixed a bug: the pen would be jagged if thicker than 4 pixels
- Fixed a bug: if the pen left the canvas it would continue to draw
- Switch to proper JSDoc
- Cleanup and removal of extraneous semi-colons

**1.5.0 (Dec. 3, 2010)**

- Added an option to disable field validation
- Cleaned up some code

**1.4.0 (Nov. 30, 2010)**

- Regenerate method now repopulates the output so signature can be appended
- Removed some extra debug code that got missed

**1.3.2 (Oct. 31, 2010)**

- Fixed a bug on iPhone created by the fix in v1.3.1: if signature pad was at the top iPhone may scroll upwards

**1.3.1 (Oct. 30, 2010)**

- Fixed a bug on iPad when zoomed in: the signature would draw in the wrong location

**1.3.0 (Aug. 29, 2010)**

- Added `getSignatureImage()` method that calls `toDataURL()` on the canvas

**1.2.2 (Aug. 29, 2010)**

- Fixed an offset bug

**1.2.1 (Jun. 7, 2010)**

- Added namespaces to all event bindings

**1.2.0 (May 28, 2010)**

- Added touch device support

**1.1.3 (May 28, 2010)**


build.js
#!/usr/bin/env node

/*
 * Module dependencies
 */
var fs = require('fs')
  , exec = require('child_process').exec
  , jshint = require('jshint').JSHINT

var jshintrc = JSON.parse(fs.readFileSync('.jshintrc', 'utf8'))
  , sigPadPath = 'jquery.signaturepad.js'
  , sigPadMinPath = 'build/jquery.signaturepad.min.js'
  , source = fs.readFileSync(sigPadPath, 'utf8')
  , valid = jshint(source, jshintrc)

process.stdout.write('JSHinting... ')

if (valid) {
  process.stdout.write('done.\nMinifying... ')

  exec('java -jar ~/bin/compiler.jar --js ' + sigPadPath + ' --js_output_file ' + sigPadMinPath
    , function(err, stdout, stderr) {
      var sp = fs.readFileSync(sigPadMinPath, 'utf8')
        , ver = fs.readFileSync('VERSION.txt', 'utf8')
        , spVer = sp.replace(/{{version}}/, ver.trim())

      fs.writeFileSync(sigPadMinPath, spVer)

      process.stdout.write('done.\n')
  })
} else {
  var e = null

  console.log('\nJSHint failed with errors:')

  for (var i = 0, t = jshint.errors.length; i < t; i++) {
    e = jshint.errors[i]
    console.log('• Line: ' + e.line + ' Col: ' + e.character + '; ' + e.reason)
  }

  console.log('Build incomplete.')
}


signaturepad.js
/*
 SignaturePad: A library for assisting in the creation of an HTML5 canvas
 based signature pad. Records the drawn signature in JSON for later regeneration.
 
 Based on the SignaturePad jQuery plugin from http://github.com/thomasjbradley/signature-pad
 Without all that jQuery nonsense
*/

(function () {

function SignaturePad (selector, options) {
  /**
   * Reference to the object for use in public methods
   *
   * @private
   *
   * @type {Object}
   */
  var self = this

  /**
   * Holds the merged default settings and user passed settings
   *
   * @private
   *
   * @type {Object}
   */
  , settings = window.signaturePadDefaultOptions

  /**
   * The current context, as passed by jQuery, of selected items
   *
   * @private
   *
   * @type {Object}
   */
  , context = selector

  /**
   * jQuery reference to the canvas element inside the signature pad
   *
   * @private
   *
   * @type {Object}
   */
  , canvas = context.getElementsByTagName('canvas')[0]

  /**
   * Dom reference to the canvas element inside the signature pad
   *
   * @private
   *
   * @type {Object}
   */
  , element = canvas

  /**
   * The drawing context for the signature canvas
   *
   * @private
   *
   * @type {Object}
   */
  , canvasContext = null

  /**
   * Holds the previous point of drawing
   * Disallows drawing over the same location to make lines more delicate
   *
   * @private
   *
   * @type {Object}
   */
  , previous = {'x': null, 'y': null, 'id': null}

  /**
   * An array holding all the points and lines to generate the signature
   * Each item is an object like:
   * {
   *   mx: moveTo x coordinate
   *   my: moveTo y coordinate
   *   lx: lineTo x coordinate
   *   lx: lineTo y coordinate
   * }
   *
   * @private
   *
   * @type {Array}
   */
  , output = []

  /**
   * Stores a timeout for when the mouse leaves the canvas
   * If the mouse has left the canvas for a specific amount of time
   * Stops drawing on the canvas
   *
   * @private
   *
   * @type {Object}
   */
  , mouseLeaveTimeout = false

  /**
   * Whether events have already been bound to the canvas or not
   *
   * @private
   *
   * @type {Boolean}
   */
  , eventsBound = false;

  for(var i in options) {
	settings[i] = options[i];
  }
  
  /** Works out the on-screen position of an element */
	function getAbsolutePosition(el) {
		var x = el.offsetLeft;
		var y = el.offsetTop;
		var offsetParent = el.offsetParent;
		
		while((el = el.parentNode) && el.scrollLeft != undefined)
		{
			x -= el.scrollLeft;
			y -= el.scrollTop;
			if (el == offsetParent)
			{
				x += el.offsetLeft;
				y += el.offsetTop;
				offsetParent = el.offsetParent;
			}
		}
		
		return {
			"x": x,
			"y": y
		}
	}

  /**
   * Draws a line on canvas using the mouse position
   * Checks previous position to not draw over top of previous drawing
   *  (makes the line really thick and poorly anti-aliased)
   *
   * @private
   *
   * @param {Object} e The event object
   * @param {Number} newYOffset A pixel value for drawing the newY, used for drawing a single dot on click
   */
  function drawLine (e, newYOffset) {
    var offset = getAbsolutePosition(e.target), newX, newY;

	e.preventDefault();
    clearTimeout(mouseLeaveTimeout);
    mouseLeaveTimeout = false;
	
	if (e.pointerId && !e.isPrimary) return;
	
	if (typeof e.targetTouches !== 'undefined') {
      newX = Math.floor(e.targetTouches[0].pageX - offset.x - window.pageXOffset);
      newY = Math.floor(e.targetTouches[0].pageY - offset.y - window.pageYOffset);
    } else {
      newX = Math.floor(e.pageX - offset.x);
      newY = Math.floor(e.pageY - offset.y);
    }

    if (previous.x === newX && previous.y === newY) {
      return true;
	}

    if (previous.x === null) {
      previous.x = newX;
	}

    if (previous.y === null) {
      previous.y = newY;
	}

    if (newYOffset) {
      newY += newYOffset;
	}
	
	if (e.intermediatePoints) {
		canvasContext.beginPath();
		canvasContext.lineCap = settings.penCap;
		canvasContext.moveTo(previous.x, previous.y);
		
		var points = e.getIntermediatePoints(canvas);
		var point = previous;
		for(var i=0; i<points.length-1; i++) {
			point = points[i];
			canvasContext.lineTo(point.x, point.y);
		}
		previous.x = point.x;
		previous.y = point.y;
		
		canvasContext.stroke();
		canvasContext.closePath();
	}

    canvasContext.beginPath();
    canvasContext.moveTo(previous.x, previous.y);
    canvasContext.lineTo(newX, newY);
    canvasContext.lineCap = settings.penCap;
    canvasContext.stroke();
    canvasContext.closePath();

    output.push({
      'lx' : newX
      , 'ly' : newY
      , 'mx' : previous.x
      , 'my' : previous.y
    });

    previous.x = newX;
    previous.y = newY;
  }

  /**
   * Callback registered to mouse/touch events of the canvas
   * Stops the drawing abilities
   *
   * @private
   *
   * @param {Object} e The event object
   */
  function stopDrawing (e) {
	if (e && e.which && e.which > 1) return;
	if (e && e.changedTouches && e.changedTouches[0].identifier != previous.id) return;
	if (e && e.pointerId && !e.isPrimary) return;
	
    canvas.removeEventListener('touchmove', drawLine);
	canvas.removeEventListener('MSPointerMove', drawLine);
    canvas.removeEventListener('mousemove', drawLine);

    previous.x = null;
    previous.y = null;
	previous.id = null;

    if (settings.output && output.length > 0) {
		document.getElementById(settings.output.replace(/^#/,'')).value = JSON.stringify(output);
	}
  }

  /**
   * Draws the signature line
   *
   * @private
   */
  function drawSigLine () {
    if (!settings.lineWidth) {
      return false;
	}

    canvasContext.beginPath();
    canvasContext.lineWidth = settings.lineWidth;
    canvasContext.strokeStyle = settings.lineColour;
    canvasContext.moveTo(settings.lineMargin, settings.lineTop);
    canvasContext.lineTo(element.width - settings.lineMargin, settings.lineTop);
    canvasContext.stroke();
    canvasContext.closePath();
  }

  /**
   * Clears all drawings off the canvas and redraws the signature line
   *
   * @private
   */
  function clearCanvas () {
    stopDrawing();

    canvasContext.clearRect(0, 0, element.width, element.height);
    canvasContext.fillStyle = settings.bgColour;
    canvasContext.fillRect(0, 0, element.width, element.height);

    if (!settings.displayOnly) {
      drawSigLine();
	}

    canvasContext.lineWidth = settings.penWidth;
    canvasContext.strokeStyle = settings.penColour;

    document.getElementById(settings.output.replace(/^#/,'')).value = '';
    output = [];
  }

  /**
   * Callback registered to mouse/touch events of canvas
   * Triggers the drawLine function
   *
   * @private
   *
   * @param {Object} e The event object
   * @param {Object} o The object context registered to the event; canvas
   */
  function startDrawing (e, o) {
	if (e.targetTouches && e.targetTouches[0] != e.changedTouches[0]) return;
  
    canvas.addEventListener('touchmove', drawLine, false);
	if (window.navigator.msPointerEnabled) {
		canvas.addEventListener('MSPointerMove', drawLine, false);
	} else {
		canvas.addEventListener('mousemove', drawLine, false);
	}
	
	if (e.targetTouches) previous.id = e.targetTouches[0].identifier;

    // Draws a single point on initial mouse down, for people with periods in their name
    drawLine(e, 1);
  }

  /**
   * Removes all the mouse events from the canvas
   *
   * @private
   */
  function disableCanvas () {
    eventsBound = false;

    canvas.removeEventListener('touchend', stopDrawing);
    canvas.removeEventListener('touchcancel', stopDrawing);
    canvas.removeEventListener('touchmove', drawLine);
    canvas.removeEventListener('touchleave', mouseOut);
    canvas.removeEventListener('touchstart', touchStart);

    canvas.removeEventListener('mousedown', mouseDown);
    window.removeEventListener('mouseup', stopDrawing);
    canvas.removeEventListener('mousemove', drawLine);
    canvas.removeEventListener('mouseout', mouseOut);

    canvas.removeEventListener('MSPointerDown', mouseDown);
    canvas.removeEventListener('MSPointerUp', stopDrawing);
    canvas.removeEventListener('MSPointerMove', drawLine);
    canvas.removeEventListener('MSPointerOut', mouseOut);
	
	if (canvas.oncontextmenu) canvas.oncontextmenu = null;
  }

  /**
   * Lazy touch event detection
   * Uses the first press on the canvas to detect either touch or mouse reliably
   * Will then bind other events as needed
   *
   * @private
   *
   * @param {Object} e The event object
   */
  function initDrawEvents (e) {
    if (eventsBound) {
      return false;
	}

    eventsBound = true;

    // Closes open keyboards to free up space
	if(document.activeElement) {
		document.activeElement.blur();
	}

    canvas.addEventListener('touchend', stopDrawing, false);
    canvas.addEventListener('touchcancel', stopDrawing, false);
    window.addEventListener('mouseup', stopDrawing, false);
    canvas.addEventListener('MSPointerUp', stopDrawing, false);
    canvas.addEventListener('MSPointerCancel', stopDrawing, false);
	
    canvas.addEventListener('MSPointerOut', mouseOut, false);
    canvas.addEventListener('mouseout', mouseOut, false);
    canvas.addEventListener('touchleave', mouseOut, false);
	
    canvas.removeEventListener('MSPointerDown', mouseDown);
    canvas.removeEventListener('mousedown', mouseDown);
    canvas.removeEventListener('touchstart', touchStart);
  }
	function mouseOut(e) {
        if (!mouseLeaveTimeout) {
          mouseLeaveTimeout = setTimeout(function () {
            stopDrawing();
            clearTimeout(mouseLeaveTimeout);
            mouseLeaveTimeout = false;
          }, 1000);
        }
		
		drawLine(e);
		
		previous.x = null;
		previous.y = null;
	}
	function mouseDown(e) {
      // Only allow left mouse clicks to trigger signature drawing
      if (e.which > 1) return false;
	  
      initDrawEvents(e);
      startDrawing(e, this);
	}
	function touchStart(e) {
		e.preventDefault();
        initDrawEvents(e);
        startDrawing(e, this);
	}
  /**
   * Triggers the abilities to draw on the canvas
   * Sets up mouse/touch events, hides and shows descriptions and sets current classes
   *
   * @private
   */
  function drawIt () {
    clearCanvas();

	canvas.addEventListener('touchstart', touchStart, true);
	if (window.navigator.msPointerEnabled) {
		canvas.addEventListener('MSPointerDown', mouseDown, true);
	} else {
		canvas.addEventListener('mousedown', mouseDown, true);
	}

  }


  /**
   * Redraws the signature on a specific canvas
   *
   * @private
   *
   * @param {Array} paths the signature JSON
   * @param {Object} context the canvas context to draw on
   * @param {Boolean} saveOutput whether to write the path to the output array or not
   */
  function drawSignature (paths, context, saveOutput) {
	for(var i in paths) {
      if (typeof paths[i] === 'object') {
        context.beginPath();
        context.moveTo(paths[i].mx, paths[i].my);
        context.lineTo(paths[i].lx, paths[i].ly);
        context.lineCap = settings.penCap;
        context.stroke();
        context.closePath();

        if (saveOutput) {
          output.push({
            'lx' : paths[i].lx
            , 'ly' : paths[i].ly
            , 'mx' : paths[i].mx
            , 'my' : paths[i].my
          });
        }
      }
    }
  }

  /**
   * Initialisation function, called immediately after all declarations
   * Technically public, but only should be used internally
   *
   * @private
   */
  function init () {

    // Disable selection on the typed div and canvas
	context.addEventListener('selectstart', function (e) { return e.target.tagName == 'INPUT'; }, true);
    canvas.addEventListener('selectstart', function (e) { return e.target.tagName == 'INPUT'; }, true);
	
	canvas.oncontextmenu = function() { return false; };

    if (!element.getContext && window.FlashCanvas) {
      FlashCanvas.initElement(element);
	}

    if (element.getContext) {
      canvasContext = element.getContext('2d');

      if (!settings.displayOnly) {
		drawIt();
      }
    }
  }

  
    /**
     * Initializes SignaturePad
     */
    self.init = function () { init(); };

    /**
     * Allows options to be updated after initialization
     *
     * @param {Object} options An object containing the options to be changed
     */
    self.updateOptions = function (options) {
      for(var i in options) {
		settings[i] = options[i];
	  }
    };

    /**
     * Regenerates a signature on the canvas using an array of objects
     * Follows same format as object property
     * @see var object
     *
     * @param {Array} paths An array of the lines and points
     */
    self.regenerate = function (paths) {
      self.clearCanvas();

      if (typeof paths === 'string') {
        paths = JSON.parse(paths);
	  }

      drawSignature(paths, canvasContext, true);

      if (settings.output) {
        document.getElementById(settings.output.replace(/^#/,'')).value = JSON.stringify(output);
	  }
    };

    /**
     * Clears the canvas
     * Redraws the background colour and the signature line
     */
    self.clearCanvas = function () { clearCanvas(); };

    /**
     * Returns the signature as a Js array
     *
     * @return {Array}
     */
    self.getSignature = function () { return output; };

    /**
     * Returns the signature as a Json string
     *
     * @return {String}
     */
    self.getSignatureString = function () { return JSON.stringify(output); };

    /**
     * Returns the signature as an image
     * Re-draws the signature in a shadow canvas to create a clean version
     *
     * @return {String}
     */
    self.getSignatureImage = function () {
      var tmpCanvas = document.createElement('canvas')
        , tmpContext = null
        , data = null;

      tmpCanvas.style.position = 'absolute';
      tmpCanvas.style.top = '-999em';
      tmpCanvas.width = element.width;
      tmpCanvas.height = element.height;
      document.body.appendChild(tmpCanvas);

      if (!tmpCanvas.getContext && window.FlashCanvas) {
        FlashCanvas.initElement(tmpCanvas);
	  }

      tmpContext = tmpCanvas.getContext('2d');

      tmpContext.fillStyle = settings.bgColour;
      tmpContext.fillRect(0, 0, element.width, element.height);
      tmpContext.lineWidth = settings.penWidth;
      tmpContext.strokeStyle = settings.penColour;

      drawSignature(output, tmpContext);
      data = tmpCanvas.toDataURL.apply(tmpCanvas, arguments);

      document.body.removeChild(tmpCanvas);
      tmpCanvas = null;

      return data;
    };
  
}

/**
 * Create the plugin
 * Returns an Api which can be used to call specific methods
 *
 * @param {Object} options The options array
 *
 * @return {Object} The Api for controlling the instance
 */
window.createSignaturePad = function (element, options) {
  var api = null;
	
	if (!element.sigPad) {
		api = new SignaturePad(element, options);
		api.init();
		element.sigPad = api;
	} else {
		api = element.sigPad;
		api.updateOptions(options);
	}

  return api;
};

/**
 * Expose the defaults so they can be overwritten for multiple instances
 *
 * @type {Object}
 */
window.signaturePadDefaultOptions = {
  displayOnly : false // Initialize canvas for signature display only; ignore buttons and inputs
  , bgColour : '#ffffff' // The colour fill for the background of the canvas; or transparent
  , penColour : '#145394' // Colour of the drawing ink
  , penWidth : 2 // Thickness of the pen
  , penCap : 'round' // Determines how the end points of each line are drawn (values: 'butt', 'round', 'square')
  , lineColour : '#ccc' // Colour of the signature line
  , lineWidth : 2 // Thickness of the signature line
  , lineMargin : 5 // Margin on right and left of signature line
  , lineTop : 35 // Distance to draw the line from the top
  , output : '.output' // The hidden input field for remembering line coordinates
};
}());


examples/sample-signature-output.js
var sig = [
{"lx":20,"ly":34,"mx":20,"my":34},
{"lx":21,"ly":33,"mx":20,"my":34},
{"lx":22,"ly":31,"mx":21,"my":33},
{"lx":23,"ly":30,"mx":22,"my":31},
{"lx":25,"ly":27,"mx":23,"my":30},
{"lx":27,"ly":25,"mx":25,"my":27},
{"lx":29,"ly":23,"mx":27,"my":25},
{"lx":31,"ly":21,"mx":29,"my":23},
{"lx":33,"ly":19,"mx":31,"my":21},
{"lx":35,"ly":18,"mx":33,"my":19},
{"lx":35,"ly":17,"mx":35,"my":18},
{"lx":36,"ly":17,"mx":35,"my":17},
{"lx":36,"ly":17,"mx":36,"my":17},
{"lx":35,"ly":18,"mx":36,"my":17},
{"lx":34,"ly":20,"mx":35,"my":18},
{"lx":32,"ly":23,"mx":34,"my":20},
{"lx":30,"ly":26,"mx":32,"my":23},
{"lx":28,"ly":29,"mx":30,"my":26},
{"lx":26,"ly":32,"mx":28,"my":29},
{"lx":24,"ly":33,"mx":26,"my":32},
{"lx":23,"ly":35,"mx":24,"my":33},
{"lx":23,"ly":36,"mx":23,"my":35},
{"lx":23,"ly":36,"mx":23,"my":36},
{"lx":24,"ly":35,"mx":23,"my":36},
{"lx":26,"ly":34,"mx":24,"my":35},
{"lx":28,"ly":33,"mx":26,"my":34},
{"lx":30,"ly":31,"mx":28,"my":33},
{"lx":32,"ly":30,"mx":30,"my":31},
{"lx":33,"ly":29,"mx":32,"my":30},
{"lx":34,"ly":29,"mx":33,"my":29},
{"lx":34,"ly":29,"mx":34,"my":29},
{"lx":33,"ly":30,"mx":34,"my":29},
{"lx":32,"ly":31,"mx":33,"my":30},
{"lx":31,"ly":33,"mx":32,"my":31},
{"lx":30,"ly":35,"mx":31,"my":33},
{"lx":29,"ly":36,"mx":30,"my":35},
{"lx":29,"ly":37,"mx":29,"my":36},
{"lx":29,"ly":37,"mx":29,"my":37},
{"lx":31,"ly":36,"mx":29,"my":37},
{"lx":33,"ly":35,"mx":31,"my":36},
{"lx":36,"ly":34,"mx":33,"my":35},
{"lx":38,"ly":32,"mx":36,"my":34},
{"lx":40,"ly":31,"mx":38,"my":32},
{"lx":42,"ly":30,"mx":40,"my":31},
{"lx":43,"ly":30,"mx":42,"my":30},
{"lx":44,"ly":30,"mx":43,"my":30},
{"lx":44,"ly":31,"mx":44,"my":30},
{"lx":44,"ly":32,"mx":44,"my":31},
{"lx":43,"ly":33,"mx":44,"my":32},
{"lx":42,"ly":35,"mx":43,"my":33},
{"lx":41,"ly":36,"mx":42,"my":35},
{"lx":40,"ly":37,"mx":41,"my":36},
{"lx":40,"ly":37,"mx":40,"my":37},
{"lx":40,"ly":36,"mx":40,"my":37},
{"lx":41,"ly":36,"mx":40,"my":36},
{"lx":42,"ly":35,"mx":41,"my":36},
{"lx":44,"ly":33,"mx":42,"my":35},
{"lx":46,"ly":32,"mx":44,"my":33},
{"lx":47,"ly":32,"mx":46,"my":32},
{"lx":48,"ly":31,"mx":47,"my":32},
{"lx":48,"ly":31,"mx":48,"my":31},
{"lx":48,"ly":33,"mx":48,"my":31},
{"lx":47,"ly":35,"mx":48,"my":33},
{"lx":45,"ly":37,"mx":47,"my":35},
{"lx":43,"ly":39,"mx":45,"my":37},
{"lx":41,"ly":42,"mx":43,"my":39},
{"lx":39,"ly":44,"mx":41,"my":42},
{"lx":37,"ly":45,"mx":39,"my":44},
{"lx":37,"ly":46,"mx":37,"my":45},
{"lx":37,"ly":46,"mx":37,"my":46},
{"lx":37,"ly":45,"mx":37,"my":46},
{"lx":39,"ly":44,"mx":37,"my":45},
{"lx":42,"ly":42,"mx":39,"my":44},
{"lx":46,"ly":40,"mx":42,"my":42},
{"lx":49,"ly":38,"mx":46,"my":40},
{"lx":53,"ly":36,"mx":49,"my":38},
{"lx":56,"ly":35,"mx":53,"my":36},
{"lx":58,"ly":34,"mx":56,"my":35},
{"lx":59,"ly":34,"mx":58,"my":34},
{"lx":60,"ly":34,"mx":59,"my":34},
{"lx":59,"ly":35,"mx":60,"my":34},
{"lx":58,"ly":36,"mx":59,"my":35},
{"lx":57,"ly":37,"mx":58,"my":36},
{"lx":55,"ly":39,"mx":57,"my":37},
{"lx":54,"ly":40,"mx":55,"my":39},
{"lx":53,"ly":41,"mx":54,"my":40},
{"lx":53,"ly":41,"mx":53,"my":41},
{"lx":55,"ly":40,"mx":53,"my":41},
{"lx":57,"ly":38,"mx":55,"my":40},
{"lx":60,"ly":36,"mx":57,"my":38},
{"lx":63,"ly":34,"mx":60,"my":36},
{"lx":65,"ly":32,"mx":63,"my":34},
{"lx":67,"ly":31,"mx":65,"my":32},
{"lx":69,"ly":31,"mx":67,"my":31},
{"lx":69,"ly":31,"mx":69,"my":31},
{"lx":68,"ly":31,"mx":69,"my":31},
{"lx":66,"ly":33,"mx":68,"my":31},
{"lx":64,"ly":34,"mx":66,"my":33},
{"lx":62,"ly":36,"mx":64,"my":34},
{"lx":60,"ly":37,"mx":62,"my":36},
{"lx":59,"ly":38,"mx":60,"my":37},
{"lx":58,"ly":39,"mx":59,"my":38},
{"lx":58,"ly":39,"mx":58,"my":39},
{"lx":59,"ly":39,"mx":58,"my":39},
{"lx":61,"ly":38,"mx":59,"my":39},
{"lx":64,"ly":36,"mx":61,"my":38},
{"lx":66,"ly":35,"mx":64,"my":36},
{"lx":69,"ly":33,"mx":66,"my":35},
{"lx":72,"ly":32,"mx":69,"my":33},
{"lx":73,"ly":31,"mx":72,"my":32},
{"lx":74,"ly":31,"mx":73,"my":31},
{"lx":74,"ly":31,"mx":74,"my":31},
{"lx":72,"ly":31,"mx":74,"my":31},
{"lx":70,"ly":32,"mx":72,"my":31},
{"lx":68,"ly":33,"mx":70,"my":32},
{"lx":65,"ly":34,"mx":68,"my":33},
{"lx":64,"ly":36,"mx":65,"my":34},
{"lx":62,"ly":36,"mx":64,"my":36},
{"lx":61,"ly":37,"mx":62,"my":36},
{"lx":61,"ly":37,"mx":61,"my":37},
{"lx":62,"ly":37,"mx":61,"my":37},
{"lx":64,"ly":36,"mx":62,"my":37},
{"lx":68,"ly":34,"mx":66,"my":35},
{"lx":69,"ly":33,"mx":68,"my":34},
{"lx":70,"ly":33,"mx":69,"my":33},
{"lx":70,"ly":33,"mx":70,"my":33},
{"lx":70,"ly":32,"mx":70,"my":33},
{"lx":70,"ly":32,"mx":70,"my":32},
{"lx":71,"ly":32,"mx":70,"my":32},
{"lx":72,"ly":32,"mx":71,"my":32},
{"lx":74,"ly":32,"mx":72,"my":32},
{"lx":76,"ly":32,"mx":74,"my":32},
{"lx":79,"ly":33,"mx":76,"my":32},
{"lx":83,"ly":33,"mx":79,"my":33},
{"lx":87,"ly":34,"mx":83,"my":33},
{"lx":92,"ly":35,"mx":87,"my":34},
{"lx":97,"ly":36,"mx":92,"my":35},
{"lx":103,"ly":37,"mx":97,"my":36},
{"lx":110,"ly":38,"mx":103,"my":37},
{"lx":117,"ly":39,"mx":110,"my":38},
{"lx":125,"ly":39,"mx":117,"my":39},
{"lx":133,"ly":38,"mx":125,"my":39},
{"lx":140,"ly":36,"mx":133,"my":38},
{"lx":146,"ly":34,"mx":140,"my":36},
{"lx":151,"ly":32,"mx":146,"my":34},
{"lx":155,"ly":30,"mx":151,"my":32},
{"lx":157,"ly":28,"mx":155,"my":30},
{"lx":158,"ly":27,"mx":157,"my":28},
{"lx":158,"ly":27,"mx":158,"my":27}
];
